dali/operators/image/distortion/jpeg_compression_distortion_op_gpu.cu

```diff
@@ -0,0 +1,109 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <vector>
+#include "dali/operators/image/distortion/jpeg_compression_distortion_op.h"
+#include "dali/kernels/imgproc/jpeg/jpeg_distortion_gpu.cuh"
+#include "dali/kernels/common/block_setup.h"
+
+namespace dali {
+
+class JpegCompressionDistortionGPU : public JpegCompressionDistortion<GPUBackend> {
+ public:
+  explicit JpegCompressionDistortionGPU(const OpSpec &spec) : JpegCompressionDistortion(spec) {}
+
+  using Operator<GPUBackend>::RunImpl;
+  ~JpegCompressionDistortionGPU() override = default;
+  DISABLE_COPY_MOVE_ASSIGN(JpegCompressionDistortionGPU);
+
+ protected:
+  void RunImpl(workspace_t<GPUBackend> &ws) override;
+
+ private:
+  using BlkSetup = kernels::BlockSetup<2, -1>;
+  BlkSetup block_setup_;
+  using BlockDesc = BlkSetup::BlockDesc;
+
+  using SampleDesc = kernels::SampleDesc;
+  std::vector<SampleDesc> samples_cpu_;
+  Tensor<GPUBackend> scratchpad_gpu_;
+};
+
+void JpegCompressionDistortionGPU::RunImpl(workspace_t<GPUBackend> &ws) {
+  const auto &input = ws.InputRef<GPUBackend>(0);
+  auto &output = ws.OutputRef<GPUBackend>(0);
+  auto in_view = view<const uint8_t>(input);
+  auto out_view = view<uint8_t>(output);
+  int nsamples = in_view.shape.size();
+
+  TensorListShape<2> chroma_shape(nsamples, 2);
+  samples_cpu_.resize(nsamples);
+
+  for (int i = 0; i < nsamples; i++) {
+    auto &sample_desc = samples_cpu_[i];
+    const auto& in_sh = in_view.shape[i];
+    auto chroma_sh = chroma_shape.tensor_shape_span(i);
+    auto shape_vol = volume(in_sh);
+    auto width = in_sh[1];
+    auto height = in_sh[0];
+    auto chroma_width = div_ceil(width, 2);
+    auto chroma_height = div_ceil(height, 2);
+    // used to generate logical blocks (one thread per chroma pixel)
+    chroma_sh[0] = chroma_height;
+    chroma_sh[1] = chroma_width;
+    chroma_sh[2] = 3;
+
+    sample_desc.in = in_view[i].data;
+    sample_desc.out = out_view[i].data;
+    sample_desc.size.x = width;
+    sample_desc.size.y = height;
+    sample_desc.strides.x = 3;
+    sample_desc.strides.y = width * 3;
+    auto q = quality_[i].data[0];
+    sample_desc.luma_Q_table = kernels::GetLumaQuantizationTable(q);
+    sample_desc.chroma_Q_table = kernels::GetChromaQuantizationTable(q);
+  }
+  block_setup_.SetBlockDim(dim3(32, 16, 1));
+  int xblock = 64;
+  int yblock = 128;
+  block_setup_.SetDefaultBlockSize({xblock, yblock});
+  block_setup_.SetupBlocks(chroma_shape, true);
+  auto blocks_cpu = block_setup_.Blocks();
+
+  int64_t sample_descs_sz = samples_cpu_.size() * sizeof(SampleDesc);
+  int64_t block_descs_sz = blocks_cpu.size() * sizeof(BlockDesc);
+  scratchpad_gpu_.set_type(TypeTable::GetTypeInfo(DALI_UINT8));
+  scratchpad_gpu_.Resize({sample_descs_sz + block_descs_sz});
```

Comment:
This code is very tightly coupled to the implementation of the kernel and should go along with it. It could either be a DALI kernel (with Setup and Run) or a function or small set of functions - I'm not choosy - but it should be in dali/kernels, avilable from the same header file as the kernel function.